packages/postgres-database/src/simcore_postgres_database/utils_projects_nodes.py

```diff
@@ -0,0 +1,215 @@
+import datetime
+import uuid
+from dataclasses import asdict, dataclass, field
+
+import sqlalchemy
+from aiopg.sa.connection import SAConnection
+from sqlalchemy import literal_column
+
+from .errors import ForeignKeyViolation, UniqueViolation
+from .models.projects_nodes import projects_nodes
+from .models.projects_to_projects_nodes import projects_to_projects_nodes
+
+
+#
+# Errors
+#
+class BaseProjectsNodesError(Exception):
+    ...
+
+
+class ProjectsNodesProjectNotFound(BaseProjectsNodesError):
+    ...
+
+
+class ProjectsNodesNodeNotFound(BaseProjectsNodesError):
+    ...
+
+
+class ProjectsNodesOperationNotAllowed(BaseProjectsNodesError):
+    ...
+
+
+class ProjectsNodesDuplicateNode(BaseProjectsNodesError):
+    ...
+
+
+@dataclass(frozen=True, slots=True, kw_only=True)
+class ProjectsNodeCreate:
+    node_id: uuid.UUID
+    required_resources: dict = field(default_factory=dict)
+
+
+@dataclass(frozen=True, slots=True, kw_only=True)
+class ProjectsNode(ProjectsNodeCreate):
+    created: datetime.datetime
+    modified: datetime.datetime
+
+
+@dataclass(frozen=True, slots=True, kw_only=True)
+class ProjectsNodesRepo:
+    project_uuid: uuid.UUID
+
+    async def create(
+        self, connection: SAConnection, *, node: ProjectsNodeCreate
```

Comment:
THOUGH: one possibilityh is to add the fields as parameters here instead of having an object that cotnains them all

Reply:
hmm, actually I do not like having 20 parameters in a function. currently this is only for resources, but on the longer term this should contain the whole workbench stuff.